snakemake/executors/__init__.py

```diff
@@ -744,11 +744,13 @@ def format_job(self, pattern, job, **kwargs):
         if self.assume_shared_fs:
             wait_for_files.append(self.tmpdir)
             wait_for_files.extend(job.get_wait_for_files())
-            # Prepend PATH of current python executable to PATH.
+            # Append PATH of current python executable to PATH.
             # This way, we ensure that the snakemake process in the cluster node runs
-            # in the same environment as the current process.
-            # This is necessary in order to find the pulp solver backends (e.g. coincbc).
-            path = "PATH='{}':$PATH".format(os.path.dirname(sys.executable))
+            # with the pulp solver backends (e.g. coincbc) available on the PATH.
+            # Appending ensures that Snakemake's Python cannot supersede Python
+            # from a job's environment (conda activate replaces the first PATH
+            # corresponding to the current environment rather than prepending)
+            path = "PATH=${{PATH}}:'{}'".format(os.path.dirname(sys.executable))
```

Comment:
Mhm. Not entirely sure about this. On, e.g., the login node, you start snakemake which submits jobs. Why should these jobs prepend an environment via conda which is something else than the one that you use on the login note to start snakemake?

Reply:
I was thinking of the case when we are using `--use-conda`, and the rule has the `conda:` directive specifying a different conda environment to make. In this case, the current approach (prepending) can lead to the wrong Python being first in the PATH.

For example:
```
1. (qsub batch job starts with activated default conda environment)
    PATH="{sge-binaries}/bin:$HOME/miniconda3/bin:{...}"
2. (Snakemake job script prepends snakemake bin directory)
    PATH="$HOME/miniconda3/envs/snakemake/bin:{sge-binaries}/bin:$HOME/miniconda3/bin:{...}"
3. (Snakemake activates desired conda environment, replacing default conda environment)
    PATH="$HOME/miniconda3/envs/snakemake/bin:{sge-binaries}/bin:{pipeline-dir}/.snakemake/conda/{env-hash}/bin:{...}"

(python runs from $HOME/miniconda3/envs/snakemake/bin, not {pipeline-dir}/.snakemake/conda/{env-hash}/bin...)
```

We want binaries (e.g. Python) from the conda rule directive to take precedence over Python from those in the Snakemake environment.

However, I interpret your comment as rightfully pointing out that we run into the reverse issue if we are not using conda and we use the naive appending solution. Then, to be similar to running locally, we effectively want to use the PATH from where Snakemake was executed. But, if we append, the version of Python we'll end up using will probably be default conda, or worse, the system Python.

A simple solution could be: append if we are using a conda environment, otherwise prepend. What do you think @johanneskoester?

Reply:
@johanneskoester does Snakemake make the assumption that the cluster environment (from `bsub`, `qsub`, `sbatch`, etc.) inherits environment variables from the node on which Snakemake itself runs?

My HPC is transitioning away from SGE (the scheduler that this problem, i.e. #883, seems to always occur on) to slurm. By default Slurm and bsub (the other scheduler I've worked with in the past) copy environment variables, while qsub does not.

The current master should work if:
```
(
 Snakemake is run from the active conda environment
 AND
 the job submission copies the environment variables (in particular those for conda) to the compute node
)
OR
The compute node starts with no conda environment activated
```

If the environment variables assumption is correct, then I suspect a lot of SGE users are breaking it (using older versions of the suggested snakemake profile or from rolling their own -- [the intro to cluster execution](https://snakemake.readthedocs.io/en/stable/executing/cluster.html) uses qsub as an example without the `-V` flag). But, then the solution would be turning this PR into (or starting a new PR to) update documentation to explicitly state this assumption?